File: TestOverload.scala
class Test {
  def foo(): String = {
    return "?"
  }
  def foo(i: Int): String = {
    if (i == 1) return "bar"
    return foo()
  }
  def foo(s: String): String = {
    if (s.equals("1")) return "baz";
    return foo()
  }
}

object Test {
  def main(args: Array[String]) {
    println(new Test().foo( 0 ));
    println(new Test().foo( 1 ));
    println(new Test().foo("0"));
    println(new Test().foo("1"));
  }
}

File: TestOverload.java
class Test {
    public static String foo() {
        return "?";
    }
    public static String foo(int i) {
        if (i == 1) return "bar";
        return foo();
    }
    public static String foo(String s) {
        if (s.equals("1")) return "baz";
        return foo();
    }
    public static void main(String[] args) {
        System.out.println(foo( 0 ));
        System.out.println(foo( 1 ));
        System.out.println(foo("0"));
        System.out.println(foo("1"));
    }
}
